### PR TITLE
PTV-1620 - fix browser freeze and crash when rendering large heatmaps

### PR DESCRIPTION
**Pull request by eapearson**
# Description of PR purpose/changes

This change fixes PTV-1620, which reflects a public ticket that cited two issues when viewing a FeatureCluster object
- a data-tables error
- browser freezing and/or crashing

The data-tables error was due to a missing field which was not supplied with a default value; data-tables does not like undefined values. This issue was fixed separately.

The changes in this PR fix the browser freezing and crashing issue. That problem is due to the creation of a heatmap when the number of features is greater than about 200. The heatmap is grows by N², where N is the number of features. The processing of the data for the heatmap beyond 200 features will begin to consume several seconds, and at a certain point may cause the browser to crash or report an unresponsive tab.

It is notable that the affected widget, kbaseExpressionPairwiseCorrelation, already restricted the display of a heatmap with over 50 features. With over 50 features in a cluster, it provides a link for downloading the heatmap, which is hidden.

However, it placed no limits on the creation of the heatmap. It is the creation of the heatmap that consumes a great deal of cpu and memory, leading to unresponsiveness and crashing.

This set of changes focuses on addressing this issue, but also:
- corrects and improves some code issues within the same file.
- fixes an unreported issue with downloading the heatmap (wherein the second and subsequent download attempts will fail.)

There are several UI and code issues that will be addressed by a subsequent PR, affecting quite a few additional files.

# Jira Ticket / Issue #

- Jira Ticket https://kbase-jira.atlassian.net/browse/PTV-1620

- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions

Unit tests have been added.

- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [-] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook
- [-] (Python) I have run Black and Flake8 on changed Python code manually or with a git precommit hook
- [-] Any dependent changes have been merged and published in downstream modules

# Updating Version and Release Notes (if applicable)

- [-] [Version has been bumped](https://semver.org/) for each release
- [x] [Release notes](/RELEASE_NOTES.md) have been updated for each release (and during the merge of feature branches)

### CODE REVIEW

**Comment by eapearson** [comment]
The code analysis failure is due to the invocation of a kbwidget using "new" -- the widget itself is not stored, so the "bug" is constructing an object without using it. Just the weirdness of kbwidget freaking out code analysis.

**Comment by eapearson** [comment]
I'll have a closer look at the failing test ... it works locally.

**Comment by eapearson** [comment]
Added ticket https://kbase-jira.atlassian.net/browse/PTV-1844 because I wanted to have screenshots of each of the widget states.

It may be reviewed directly by building off this PR or the branch and viewing a FeatureClusters object, e.g. as output from "Cluster Expression Data - K-Means"

File: test/unit/spec/util/asyncTools.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    const TRY_LOOP_INTERVAL = 100;
+
+    /**
+     * Returns a promise which resolves when the given duration, in milliseconds,
+     * elapses.
+     *
+     * @param {int} duration
+     * @returns {Promise<void>}
+     */
+    function waitFor(duration) {
```
Comment (briehl) [inline test/unit/spec/util/asyncTools.js:18]: This is effectively a duplicate of https://github.com/kbase/narrative/blob/develop/test/unit/testUtil.js#L63 and should just use that one.

The `tryFor` function could probably live in `testUtil` as well. At the least, this file should be moved out of the `test/spec/util` directory and up into the  should be `test` directory, as it's not a test spec.

File: kbase-extension/static/kbase/js/widgets/function_output/kbaseExpressionPairwiseCorrelation.js [hunk 2]
```diff
@@ -53,86 +94,165 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseExpressionGenesetBaseWidget', '
...
+                const url = URL.createObjectURL(file);
+                link.href = url;
+                link.download = 'pairwise.svg';
+                document.body.appendChild(link);
+                link.click();
+
+                // Theoretically this ensures that the link and url are not needed any more.
+                // I don't know of documentation for this behavior, but it does seem to work.
+                // Note that there is nothing really to "download", as the heatmap content is
+                // already loaded in the browser.
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/function_output/kbaseExpressionPairwiseCorrelation.js:116]: 👍

File: kbase-extension/static/kbase/js/widgets/function_output/kbaseExpressionPairwiseCorrelation.js [hunk 3]
```diff
@@ -53,86 +94,165 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseExpressionGenesetBaseWidget', '
...
+                setTimeout(() => {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/function_output/kbaseExpressionPairwiseCorrelation.js:117]: Is the `setTimeout` wrapper bit necessary? Is that something to do with the Blob lifecycle?

**Review by briehl — CHANGES_REQUESTED** [review]
Just some minor test file tweaks, but looks great. Thanks for doing this!

File: kbase-extension/static/kbase/js/widgets/function_output/kbaseExpressionPairwiseCorrelation.js (hunk 3 above)
Comment (eapearson) [reply kbase-extension/static/kbase/js/widgets/function_output/kbaseExpressionPairwiseCorrelation.js:117]: Just added this to the comment:

I didn't find documentation for this, but I figure that the click is
not handled until the next turn of the wheel in the JS event loop. So the
timeout just lets the event loop handle the click, and timers are executed
after any pending events, so this should ensure that the link being clicked
is not removed until the link 'click' is handled.

File: test/unit/spec/util/asyncTools.js (hunk 1 above)
Comment (eapearson) [reply test/unit/spec/util/asyncTools.js:18]: Yeah, I think this was done a long time ago in the original PR, and I was trying to make as few changes from that as possible. E.g. there was nothing else in the util directory then, just mwsTool.js!
Knowing me, I also wanted the tests to be low-dependency, and it seemed like there as a lot going on in testUtils...
Anyway, I'm happy to move the code out and into testUtils (I had moved the non-spec files out myself during testing so I could isolate tests w/o the util specs.)

Comment (eapearson) [reply test/unit/spec/util/asyncTools.js:18]: Okay, have the changes locally, will update the PR a bit later this afternoon.

Comment (eapearson) [reply test/unit/spec/util/asyncTools.js]: So the changes are up. There were changes to a dozen or so files to enable this, and I included some commented-out test config features that I've found useful for focusing tests and reducing test output clutter.
The mswUtils was moved into a unit/utils directory to clarify purpose and ease inclusion of test utilities (e.g. don't need to individually add that file, don't need to individually map it for requires). asyncUtils was removed and the non-duplicate function (tryFor) moved into testUtils.

**Review by briehl — APPROVED** [review]

**Comment by eapearson** [comment]
Just adding a comment to note that this implements a sub ticket of PTV-1620: PTV-1844

**Pull request merged**